File: Algos/DotnetAlgos/BSTs/Bst.cs
namespace DotnetAlgos.Bsts
{
    public class Bst
    {
        public BstNode root;
        public Bst()
        {
            this.root = null;
        }
        public void add(int value)
        {
            if(this.root == null)
            {
                this.root = new BstNode(value);
                return;
            }
            this.add(value, this.root);
        }
        private void add(int value, BstNode node)
        {
            // check right
            if(value >= node.value)
            {
                if(node.right == null)
                {
                    node.right = new BstNode(value);
                    return;
                }
                this.add(value, node.right);
            }
            // check left
            else
            {
                if(node.left == null)
                {
                    node.left = new BstNode(value);
                    return;
                }
                this.add(value, node.left);
            }
        }
    }
}

File: Algos/JSAlgos/bst.js
class BSTNode {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}
class BST {
    constructor() {
        this.root = null;
    }
    add(value, node = this.root) {
        if(this.root == null) {
            this.root = new BSTNode(value);
            return;
        }
        
        // check right
        if(value >= node.value) {
            if(node.right == null) {
                node.right = new BSTNode(value);
                return;
            }
            return this.add(value, node.right);
        }
        else { // check left
            if(node.left == null) {
                node.left = new BSTNode(value);
                return;
            }
            return this.add(value, node.left)
        }
    }
}   

let tre = new BST();
tre.add(5);
tre.add(15);
tre.add(10);
tre.add(1);
